fish-rust/src/trace.rs

```diff
@@ -0,0 +1,96 @@
+use crate::{
+    common::{escape_string, EscapeFlags},
+    ffi::{self, parser_t, wcharz_t},
+    wchar::L,
+    wchar_ffi::{WCharFromFFI, WCharToFFI},
+};
+use cxx::CxxWString;
+use std::{ptr, sync::atomic};
+
+#[cxx::bridge]
+mod ffi2 {
+    extern "C++" {
+        include!("wutil.h");
+        include!("parser.h");
+        type wcharz_t = super::wcharz_t;
+        type parser_t = super::parser_t;
+    }
+
+    extern "Rust" {
+        fn trace_set_enabled(do_enable: bool);
+        fn trace_enabled(parser: &parser_t) -> bool;
+        fn trace_argv(parser: &parser_t, command: wcharz_t, argv: *const CxxWString, argc: usize);
+        fn trace_if_enabled(parser: &parser_t, command: wcharz_t);
+        #[cxx_name = "trace_if_enabled"]
+        fn trace_if_enabled_args(
+            parser: &parser_t,
+            command: wcharz_t,
+            argv: *const CxxWString,
+            argc: usize,
+        );
+    }
+}
+
+static do_trace: atomic::AtomicBool = atomic::AtomicBool::new(false);
+
+fn trace_set_enabled(do_enable: bool) {
+    do_trace.store(do_enable, atomic::Ordering::Relaxed);
+}
+
+fn trace_enabled(parser: &parser_t) -> bool {
+    let ld = parser.ffi_libdata_pod_const();
+    if ld.suppress_fish_trace {
+        return false;
+    }
+    return do_trace.load(atomic::Ordering::Relaxed);
+}
+
+/// Trace an "argv": a list of arguments where the first is the command.
+fn trace_argv(parser: &parser_t, command: wcharz_t, argv: *const CxxWString, argc: usize) {
```

Comment:
The types `wcharz_t` and `CxxWString` should be confined to the FFI; we don't want these to leak into non-FFI functions because it will make them difficult to call from Rust.

Please factor this into two functions:

1. `trace_argv` as you've written, except it takes `command: &wstr` and `argv: &[&wstr]`
2. A bridge function which takes FFI types, and calls `trace_argv`.

See the functions `run_builtin ` (non-FFI) and `rust_run_builtin` (FFI) for an example of how to do this.

The bridge function should probably take `command: &wcharz_t` and `argv: &Vec<wcharz_t>`; see `rust_run_builtin` for an example of how to pass these.

Reply:
Makes sense 👍